File: oauth2/typescript/src/token/storage.ts
/**
 * Token Storage
 *
 * Pluggable token persistence interface and implementations.
 */

import { StoredTokens } from "../types";
import { StorageError } from "../error";
import * as fs from "fs/promises";
import * as path from "path";

/**
 * Token storage interface (for dependency injection).
 */
export interface TokenStorage {
  /**
   * Store tokens with key.
   */
  store(key: string, tokens: StoredTokens): Promise<void>;

  /**
   * Get tokens by key.
   */
  get(key: string): Promise<StoredTokens | null>;

  /**
   * Delete tokens by key.
   */
  delete(key: string): Promise<void>;

  /**
   * Check if tokens exist for key.
   */
  exists(key: string): Promise<boolean>;

  /**
   * List all stored keys.
   */
  listKeys(): Promise<string[]>;

  /**
   * Clear all stored tokens.
   */
  clear(): Promise<void>;
}

/**
 * In-memory token storage implementation.
 */
export class InMemoryTokenStorage implements TokenStorage {
  private tokens: Map<string, StoredTokens> = new Map();

  async store(key: string, tokens: StoredTokens): Promise<void> {
    this.tokens.set(key, tokens);
  }

  async get(key: string): Promise<StoredTokens | null> {
    return this.tokens.get(key) ?? null;
  }

  async delete(key: string): Promise<void> {
    this.tokens.delete(key);
  }

  async exists(key: string): Promise<boolean> {
    return this.tokens.has(key);
  }

  async listKeys(): Promise<string[]> {
    return Array.from(this.tokens.keys());
  }

  async clear(): Promise<void> {
    this.tokens.clear();
  }
}

/**
 * Serializable token data for file storage.
 */
interface SerializedTokens {
  accessToken: string;
  tokenType: string;
  expiresAt?: string;
  refreshToken?: string;
  scopes: string[];
  idToken?: string;
  storedAt: string;
  metadata?: Record<string, string>;
}

/**
 * File-based token storage implementation.
 */
export class FileTokenStorage implements TokenStorage {
  private directory: string;
  private fileExtension: string = ".token.json";

  constructor(directory: string) {
    this.directory = directory;
  }

  async store(key: string, tokens: StoredTokens): Promise<void> {
    await this.ensureDirectory();

    const serialized: SerializedTokens = {
      accessToken: tokens.accessToken.expose(),
      tokenType: tokens.tokenType,
      expiresAt: tokens.expiresAt?.toISOString(),
      refreshToken: tokens.refreshToken?.expose(),
      scopes: tokens.scopes,
      idToken: tokens.idToken?.expose(),
      storedAt: tokens.storedAt.toISOString(),
      metadata: tokens.metadata,
    };

    const filePath = this.getFilePath(key);
    try {
      await fs.writeFile(filePath, JSON.stringify(serialized, null, 2));
    } catch (error) {
      throw new StorageError(
        `Failed to write token file: ${error}`,
        "WriteFailed"
      );
    }
  }

  async get(key: string): Promise<StoredTokens | null> {
    const filePath = this.getFilePath(key);

    try {
      const content = await fs.readFile(filePath, "utf-8");
      const serialized: SerializedTokens = JSON.parse(content);

      // Import SecretString dynamically to avoid circular dependency
      const { SecretString } = await import("../types/token");

      return {
        accessToken: new SecretString(serialized.accessToken),
        tokenType: serialized.tokenType,
        expiresAt: serialized.expiresAt
          ? new Date(serialized.expiresAt)
          : undefined,
        refreshToken: serialized.refreshToken
          ? new SecretString(serialized.refreshToken)
          : undefined,
        scopes: serialized.scopes,
        idToken: serialized.idToken
          ? new SecretString(serialized.idToken)
          : undefined,
        storedAt: new Date(serialized.storedAt),
        metadata: serialized.metadata,
      };
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        return null;
      }
      throw new StorageError(
        `Failed to read token file: ${error}`,
        "ReadFailed"
      );
    }
  }

  async delete(key: string): Promise<void> {
    const filePath = this.getFilePath(key);
    try {
      await fs.unlink(filePath);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "ENOENT") {
        throw new StorageError(
          `Failed to delete token file: ${error}`,
          "DeleteFailed"
        );
      }
    }
  }

  async exists(key: string): Promise<boolean> {
    const filePath = this.getFilePath(key);
    try {
      await fs.access(filePath);
      return true;
    } catch {
      return false;
    }
  }

  async listKeys(): Promise<string[]> {
    try {
      await this.ensureDirectory();
      const files = await fs.readdir(this.directory);
      return files
        .filter((f) => f.endsWith(this.fileExtension))
        .map((f) => f.slice(0, -this.fileExtension.length));
    } catch {
      return [];
    }
  }

  async clear(): Promise<void> {
    const keys = await this.listKeys();
    await Promise.all(keys.map((key) => this.delete(key)));
  }

  private getFilePath(key: string): string {
    // Sanitize key for filesystem
    const sanitizedKey = this.sanitizeKey(key);
    return path.join(this.directory, sanitizedKey + this.fileExtension);
  }

  private sanitizeKey(key: string): string {
    return key
      .replace(/[/\\:*?"<>|]/g, "_")
      .slice(0, 200);
  }

  private async ensureDirectory(): Promise<void> {
    try {
      await fs.mkdir(this.directory, { recursive: true });
    } catch (error) {
      throw new StorageError(
        `Failed to create token directory: ${error}`,
        "WriteFailed"
      );
    }
  }
}

/**
 * Mock token storage for testing.
 */
export class MockTokenStorage implements TokenStorage {
  private tokens: Map<string, StoredTokens> = new Map();
  private storeHistory: Array<{ key: string; tokens: StoredTokens }> = [];
  private getHistory: string[] = [];
  private deleteHistory: string[] = [];
  private nextError?: StorageError;

  /**
   * Set the next error to throw.
   */
  setNextError(error: StorageError): this {
    this.nextError = error;
    return this;
  }

  /**
   * Pre-populate tokens.
   */
  setTokens(key: string, tokens: StoredTokens): this {
    this.tokens.set(key, tokens);
    return this;
  }

  /**
   * Get store history.
   */
  getStoreHistory(): Array<{ key: string; tokens: StoredTokens }> {
    return [...this.storeHistory];
  }

  /**
   * Get get history.
   */
  getGetHistory(): string[] {
    return [...this.getHistory];
  }

  /**
   * Get delete history.
   */
  getDeleteHistory(): string[] {
    return [...this.deleteHistory];
  }

  /**
   * Assert tokens were stored.
   */
  assertStored(key: string): void {
    const stored = this.storeHistory.find((h) => h.key === key);
    if (!stored) {
      throw new Error(`Expected tokens to be stored for key: ${key}`);
    }
  }

  async store(key: string, tokens: StoredTokens): Promise<void> {
    if (this.nextError) {
      const error = this.nextError;
      this.nextError = undefined;
      throw error;
    }
    this.storeHistory.push({ key, tokens });
    this.tokens.set(key, tokens);
  }

  async get(key: string): Promise<StoredTokens | null> {
    this.getHistory.push(key);
    if (this.nextError) {
      const error = this.nextError;
      this.nextError = undefined;
      throw error;
    }
    return this.tokens.get(key) ?? null;
  }

  async delete(key: string): Promise<void> {
    this.deleteHistory.push(key);
    if (this.nextError) {
      const error = this.nextError;
      this.nextError = undefined;
      throw error;
    }
    this.tokens.delete(key);
  }

  async exists(key: string): Promise<boolean> {
    return this.tokens.has(key);
  }

  async listKeys(): Promise<string[]> {
    return Array.from(this.tokens.keys());
  }

  async clear(): Promise<void> {
    this.tokens.clear();
  }
}

/**
 * Create in-memory token storage.
 */
export function createInMemoryStorage(): TokenStorage {
  return new InMemoryTokenStorage();
}

/**
 * Create file-based token storage.
 */
export function createFileStorage(directory: string): TokenStorage {
  return new FileTokenStorage(directory);
}

/**
 * Create mock token storage for testing.
 */
export function createMockTokenStorage(): MockTokenStorage {
  return new MockTokenStorage();
}

File: oauth2/rust/src/token/storage.rs
//! Token Storage
//!
//! Secure token storage implementations with encryption and expiration tracking.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use crate::error::{OAuth2Error, StorageError};
use crate::types::StoredTokens;

/// Token storage interface.
#[async_trait]
pub trait TokenStorage: Send + Sync {
    /// Store tokens for a key.
    async fn store(&self, key: &str, tokens: StoredTokens) -> Result<(), OAuth2Error>;

    /// Retrieve tokens for a key.
    async fn retrieve(&self, key: &str) -> Result<Option<StoredTokens>, OAuth2Error>;

    /// Delete tokens for a key.
    async fn delete(&self, key: &str) -> Result<bool, OAuth2Error>;

    /// Check if tokens exist for a key.
    async fn exists(&self, key: &str) -> Result<bool, OAuth2Error>;

    /// List all stored token keys.
    async fn list_keys(&self) -> Result<Vec<String>, OAuth2Error>;

    /// Clear all stored tokens.
    async fn clear(&self) -> Result<(), OAuth2Error>;

    /// Clear expired tokens.
    async fn clear_expired(&self) -> Result<u32, OAuth2Error>;
}

/// In-memory token storage implementation.
pub struct InMemoryTokenStorage {
    tokens: Mutex<HashMap<String, StoredTokens>>,
}

impl InMemoryTokenStorage {
    /// Create new in-memory token storage.
    pub fn new() -> Self {
        Self {
            tokens: Mutex::new(HashMap::new()),
        }
    }

    fn now_ms() -> u64 {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64
    }
}

impl Default for InMemoryTokenStorage {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl TokenStorage for InMemoryTokenStorage {
    async fn store(&self, key: &str, tokens: StoredTokens) -> Result<(), OAuth2Error> {
        self.tokens
            .lock()
            .unwrap()
            .insert(key.to_string(), tokens);
        Ok(())
    }

    async fn retrieve(&self, key: &str) -> Result<Option<StoredTokens>, OAuth2Error> {
        let tokens = self.tokens.lock().unwrap();
        Ok(tokens.get(key).cloned())
    }

    async fn delete(&self, key: &str) -> Result<bool, OAuth2Error> {
        let mut tokens = self.tokens.lock().unwrap();
        Ok(tokens.remove(key).is_some())
    }

    async fn exists(&self, key: &str) -> Result<bool, OAuth2Error> {
        let tokens = self.tokens.lock().unwrap();
        Ok(tokens.contains_key(key))
    }

    async fn list_keys(&self) -> Result<Vec<String>, OAuth2Error> {
        let tokens = self.tokens.lock().unwrap();
        Ok(tokens.keys().cloned().collect())
    }

    async fn clear(&self) -> Result<(), OAuth2Error> {
        self.tokens.lock().unwrap().clear();
        Ok(())
    }

    async fn clear_expired(&self) -> Result<u32, OAuth2Error> {
        let mut tokens = self.tokens.lock().unwrap();
        let now = Self::now_ms();
        let initial_count = tokens.len();

        tokens.retain(|_, stored| {
            stored
                .access_token_expires_at
                .map(|exp| exp > now)
                .unwrap_or(true)
        });

        Ok((initial_count - tokens.len()) as u32)
    }
}

/// Encrypted token storage wrapper.
pub struct EncryptedTokenStorage<S: TokenStorage> {
    inner: S,
    /// Encryption key (in production, use proper key management)
    _encryption_key: Vec<u8>,
}

impl<S: TokenStorage> EncryptedTokenStorage<S> {
    /// Create new encrypted token storage.
    pub fn new(inner: S, encryption_key: Vec<u8>) -> Self {
        Self {
            inner,
            _encryption_key: encryption_key,
        }
    }

    // Note: In a production implementation, these would use proper encryption
    // libraries like ring or sodiumoxide
    fn _encrypt(&self, _data: &[u8]) -> Vec<u8> {
        // Placeholder - implement proper AES-GCM encryption
        todo!("Implement proper encryption")
    }

    fn _decrypt(&self, _data: &[u8]) -> Result<Vec<u8>, OAuth2Error> {
        // Placeholder - implement proper AES-GCM decryption
        todo!("Implement proper decryption")
    }
}

#[async_trait]
impl<S: TokenStorage> TokenStorage for EncryptedTokenStorage<S> {
    async fn store(&self, key: &str, tokens: StoredTokens) -> Result<(), OAuth2Error> {
        // In production: encrypt tokens before storing
        self.inner.store(key, tokens).await
    }

    async fn retrieve(&self, key: &str) -> Result<Option<StoredTokens>, OAuth2Error> {
        // In production: decrypt tokens after retrieving
        self.inner.retrieve(key).await
    }

    async fn delete(&self, key: &str) -> Result<bool, OAuth2Error> {
        self.inner.delete(key).await
    }

    async fn exists(&self, key: &str) -> Result<bool, OAuth2Error> {
        self.inner.exists(key).await
    }

    async fn list_keys(&self) -> Result<Vec<String>, OAuth2Error> {
        self.inner.list_keys().await
    }

    async fn clear(&self) -> Result<(), OAuth2Error> {
        self.inner.clear().await
    }

    async fn clear_expired(&self) -> Result<u32, OAuth2Error> {
        self.inner.clear_expired().await
    }
}

/// Mock token storage for testing.
#[derive(Default)]
pub struct MockTokenStorage {
    tokens: Mutex<HashMap<String, StoredTokens>>,
    store_history: Mutex<Vec<(String, StoredTokens)>>,
    retrieve_history: Mutex<Vec<String>>,
    delete_history: Mutex<Vec<String>>,
    next_error: Mutex<Option<OAuth2Error>>,
    should_fail: Mutex<bool>,
}

impl MockTokenStorage {
    /// Create new mock token storage.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set next error to return.
    pub fn set_next_error(&self, error: OAuth2Error) -> &Self {
        *self.next_error.lock().unwrap() = Some(error);
        self
    }

    /// Set storage to fail all operations.
    pub fn set_should_fail(&self, should_fail: bool) -> &Self {
        *self.should_fail.lock().unwrap() = should_fail;
        self
    }

    /// Pre-populate tokens.
    pub fn add_tokens(&self, key: &str, tokens: StoredTokens) -> &Self {
        self.tokens.lock().unwrap().insert(key.to_string(), tokens);
        self
    }

    /// Get store history.
    pub fn get_store_history(&self) -> Vec<(String, StoredTokens)> {
        self.store_history.lock().unwrap().clone()
    }

    /// Get retrieve history.
    pub fn get_retrieve_history(&self) -> Vec<String> {
        self.retrieve_history.lock().unwrap().clone()
    }

    /// Get delete history.
    pub fn get_delete_history(&self) -> Vec<String> {
        self.delete_history.lock().unwrap().clone()
    }

    fn check_error(&self) -> Result<(), OAuth2Error> {
        if *self.should_fail.lock().unwrap() {
            return Err(OAuth2Error::Storage(StorageError::StorageError {
                message: "Mock storage failure".to_string(),
            }));
        }

        if let Some(error) = self.next_error.lock().unwrap().take() {
            return Err(error);
        }

        Ok(())
    }
}

#[async_trait]
impl TokenStorage for MockTokenStorage {
    async fn store(&self, key: &str, tokens: StoredTokens) -> Result<(), OAuth2Error> {
        self.check_error()?;

        self.store_history
            .lock()
            .unwrap()
            .push((key.to_string(), tokens.clone()));
        self.tokens.lock().unwrap().insert(key.to_string(), tokens);
        Ok(())
    }

    async fn retrieve(&self, key: &str) -> Result<Option<StoredTokens>, OAuth2Error> {
        self.check_error()?;

        self.retrieve_history
            .lock()
            .unwrap()
            .push(key.to_string());
        Ok(self.tokens.lock().unwrap().get(key).cloned())
    }

    async fn delete(&self, key: &str) -> Result<bool, OAuth2Error> {
        self.check_error()?;

        self.delete_history.lock().unwrap().push(key.to_string());
        Ok(self.tokens.lock().unwrap().remove(key).is_some())
    }

    async fn exists(&self, key: &str) -> Result<bool, OAuth2Error> {
        self.check_error()?;
        Ok(self.tokens.lock().unwrap().contains_key(key))
    }

    async fn list_keys(&self) -> Result<Vec<String>, OAuth2Error> {
        self.check_error()?;
        Ok(self.tokens.lock().unwrap().keys().cloned().collect())
    }

    async fn clear(&self) -> Result<(), OAuth2Error> {
        self.check_error()?;
        self.tokens.lock().unwrap().clear();
        Ok(())
    }

    async fn clear_expired(&self) -> Result<u32, OAuth2Error> {
        self.check_error()?;

        let mut tokens = self.tokens.lock().unwrap();
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;

        let initial_count = tokens.len();
        tokens.retain(|_, stored| {
            stored
                .access_token_expires_at
                .map(|exp| exp > now)
                .unwrap_or(true)
        });

        Ok((initial_count - tokens.len()) as u32)
    }
}

/// Create in-memory token storage.
pub fn create_in_memory_token_storage() -> impl TokenStorage {
    InMemoryTokenStorage::new()
}

/// Create mock token storage for testing.
pub fn create_mock_token_storage() -> MockTokenStorage {
    MockTokenStorage::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_test_tokens() -> StoredTokens {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        StoredTokens {
            access_token: "test-access-token".to_string(),
            token_type: "Bearer".to_string(),
            refresh_token: Some("test-refresh-token".to_string()),
            access_token_expires_at: Some(now + 3600000), // 1 hour from now
            refresh_token_expires_at: None,
            scope: Some("openid profile".to_string()),
            id_token: None,
            created_at: now,
            updated_at: now,
        }
    }

    #[tokio::test]
    async fn test_in_memory_store_and_retrieve() {
        let storage = InMemoryTokenStorage::new();
        let tokens = create_test_tokens();

        storage.store("user1", tokens.clone()).await.unwrap();

        let retrieved = storage.retrieve("user1").await.unwrap();
        assert!(retrieved.is_some());
        assert_eq!(retrieved.unwrap().access_token, "test-access-token");
    }

    #[tokio::test]
    async fn test_in_memory_delete() {
        let storage = InMemoryTokenStorage::new();
        let tokens = create_test_tokens();

        storage.store("user1", tokens).await.unwrap();
        assert!(storage.exists("user1").await.unwrap());

        let deleted = storage.delete("user1").await.unwrap();
        assert!(deleted);
        assert!(!storage.exists("user1").await.unwrap());
    }

    #[tokio::test]
    async fn test_in_memory_list_keys() {
        let storage = InMemoryTokenStorage::new();

        storage.store("user1", create_test_tokens()).await.unwrap();
        storage.store("user2", create_test_tokens()).await.unwrap();

        let keys = storage.list_keys().await.unwrap();
        assert_eq!(keys.len(), 2);
        assert!(keys.contains(&"user1".to_string()));
        assert!(keys.contains(&"user2".to_string()));
    }

    #[tokio::test]
    async fn test_mock_storage() {
        let storage = MockTokenStorage::new();
        let tokens = create_test_tokens();

        storage.store("test-key", tokens).await.unwrap();

        let history = storage.get_store_history();
        assert_eq!(history.len(), 1);
        assert_eq!(history[0].0, "test-key");

        let retrieved = storage.retrieve("test-key").await.unwrap();
        assert!(retrieved.is_some());

        let retrieve_history = storage.get_retrieve_history();
        assert_eq!(retrieve_history.len(), 1);
    }

    #[tokio::test]
    async fn test_mock_storage_failure() {
        let storage = MockTokenStorage::new();
        storage.set_should_fail(true);

        let result = storage.store("key", create_test_tokens()).await;
        assert!(result.is_err());
    }
}
